File: src/accessiweather/ui/dialogs/alert_dialog.py
"""Alert dialog for displaying weather alert details using wxPython."""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

import wx

from ...display.presentation.formatters import format_datetime

if TYPE_CHECKING:
    pass

logger = logging.getLogger(__name__)


def show_alert_dialog(parent, alert, settings=None) -> None:
    """
    Show the alert details dialog.

    Args:
        parent: Parent window
        alert: Weather alert object
        settings: Optional application settings (for combined-mode dispatch)

    """
    try:
        parent_ctrl = parent

        dlg = AlertDialog(parent_ctrl, alert, settings)
        dlg.ShowModal()
        dlg.Destroy()

    except Exception as e:
        logger.error(f"Failed to show alert dialog: {e}")
        wx.MessageBox(
            f"Failed to open alert details: {e}",
            "Error",
            wx.OK | wx.ICON_ERROR,
        )


class AlertDialog(wx.Dialog):
    """Dialog for displaying weather alert details."""

    def __init__(self, parent, alert, settings=None):
        """
        Initialize the alert dialog.

        Args:
            parent: Parent window
            alert: Weather alert object
            settings: Optional application settings (for combined-mode dispatch)

        """
        event = getattr(alert, "event", "Weather Alert")
        super().__init__(
            parent,
            title=f"Alert: {event}",
            size=(700, 500),
            style=wx.DEFAULT_DIALOG_STYLE | wx.RESIZE_BORDER,
        )

        self.alert = alert
        self.settings = settings

        self._create_ui()
        self._setup_accessibility()
        self.Bind(wx.EVT_CHAR_HOOK, self._on_key)

    def _create_ui(self):
        """Create the dialog UI, dispatching to separate or combined mode."""
        panel = wx.Panel(self)
        main_sizer = wx.BoxSizer(wx.VERTICAL)

        style = (
            getattr(self.settings, "alert_display_style", "separate")
            if self.settings is not None
            else "separate"
        )
        if style == "combined":
            self._create_combined_ui(panel, main_sizer)
        else:
            self._create_separate_ui(panel, main_sizer)

        panel.SetSizer(main_sizer)
        self._focus_target.SetFocus()

    def _create_combined_ui(self, panel, main_sizer):
        """Create a single TextCtrl containing the full alert, with a Close button."""
        label = wx.StaticText(panel, label="Alert:")
        label.SetFont(label.GetFont().Bold())
        main_sizer.Add(label, 0, wx.LEFT | wx.RIGHT | wx.TOP, 15)

        text = self._build_combined_text(self.alert, self.settings)
        self.combined_ctrl = wx.TextCtrl(
            panel,
            value=text,
            style=wx.TE_MULTILINE | wx.TE_READONLY | wx.TE_RICH2,
        )
        main_sizer.Add(self.combined_ctrl, 1, wx.EXPAND | wx.LEFT | wx.RIGHT | wx.BOTTOM, 15)

        self._add_action_buttons(panel, main_sizer)

        self._focus_target = self.combined_ctrl

    def _create_separate_ui(self, panel, main_sizer):
        """Build the classic separate-field UI into the provided panel/sizer."""
        # Subject field (headline + times) - gets initial focus
        subject_text = self._build_subject_text()
        subject_label = wx.StaticText(panel, label="Subject:")
        subject_label.SetFont(subject_label.GetFont().Bold())
        main_sizer.Add(subject_label, 0, wx.LEFT | wx.RIGHT | wx.TOP, 15)

        self.subject_ctrl = wx.TextCtrl(
            panel,
            value=subject_text,
            style=wx.TE_MULTILINE | wx.TE_READONLY | wx.TE_RICH2,
            size=(-1, 60),
        )
        main_sizer.Add(self.subject_ctrl, 0, wx.EXPAND | wx.LEFT | wx.RIGHT | wx.BOTTOM, 15)

        # Alert info field (severity, urgency, certainty, areas)
        info_text = self._build_info_text()
        if info_text:
            info_label = wx.StaticText(panel, label="Alert Info:")
            info_label.SetFont(info_label.GetFont().Bold())
            main_sizer.Add(info_label, 0, wx.LEFT | wx.RIGHT, 15)

            self.info_ctrl = wx.TextCtrl(
                panel,
                value=info_text,
                style=wx.TE_MULTILINE | wx.TE_READONLY | wx.TE_RICH2,
                size=(-1, 60),
            )
            main_sizer.Add(self.info_ctrl, 0, wx.EXPAND | wx.LEFT | wx.RIGHT | wx.BOTTOM, 15)

        # Details field (description)
        description = getattr(self.alert, "description", None)
        if description:
            details_label = wx.StaticText(panel, label="Details:")
            details_label.SetFont(details_label.GetFont().Bold())
            main_sizer.Add(details_label, 0, wx.LEFT | wx.RIGHT, 15)

            self.details_ctrl = wx.TextCtrl(
                panel,
                value=description,
                style=wx.TE_MULTILINE | wx.TE_READONLY | wx.TE_RICH2,
            )
            main_sizer.Add(self.details_ctrl, 1, wx.EXPAND | wx.LEFT | wx.RIGHT | wx.BOTTOM, 15)

        # Instructions field
        instruction = getattr(self.alert, "instruction", None)
        if instruction:
            instr_label = wx.StaticText(panel, label="Instructions:")
            instr_label.SetFont(instr_label.GetFont().Bold())
            main_sizer.Add(instr_label, 0, wx.LEFT | wx.RIGHT, 15)

            self.instr_ctrl = wx.TextCtrl(
                panel,
                value=instruction,
                style=wx.TE_MULTILINE | wx.TE_READONLY | wx.TE_RICH2,
                size=(-1, 80),
            )
            main_sizer.Add(self.instr_ctrl, 0, wx.EXPAND | wx.LEFT | wx.RIGHT | wx.BOTTOM, 15)

        self._add_action_buttons(panel, main_sizer)

        # Set initial focus to subject field
        self._focus_target = self.subject_ctrl

    def _add_action_buttons(self, panel, main_sizer):
        """
        Add the right-aligned Close button to the provided sizer.

        Shared by both display modes. The button row is constructed once here
        so that future additions (e.g. a Copy button) only need one change site.
        """
        button_sizer = wx.BoxSizer(wx.HORIZONTAL)
        button_sizer.AddStretchSpacer()

        close_btn = wx.Button(panel, wx.ID_CLOSE, "&Close")
        close_btn.Bind(wx.EVT_BUTTON, self._on_close)
        button_sizer.Add(close_btn, 0)

        main_sizer.Add(button_sizer, 0, wx.EXPAND | wx.ALL, 15)

    def _build_subject_text(self) -> str:
        """
        Build the subject text with headline.

        Note: Time information (effective/expires) is intentionally omitted
        as the headline already contains human-readable timing information.
        The API's 'expires' field refers to message validity, not when the
        weather event ends, which would be confusing to display.
        """
        # Headline or event name
        headline = getattr(self.alert, "headline", None)
        event = getattr(self.alert, "event", None)

        if headline:
            return headline
        if event:
            return event
        return "Weather Alert"

    @staticmethod
    def _build_combined_text(alert, settings) -> str:
        """Assemble the combined-view text block. Pure function; settings is AppSettings-like."""
        date_style = getattr(settings, "date_format", "iso")
        time_12h = getattr(settings, "time_format_12hour", True)

        blocks: list[str] = []

        headline = (
            getattr(alert, "headline", None) or getattr(alert, "event", None) or "Weather Alert"
        )
        blocks.append(headline)

        description = getattr(alert, "description", None)
        if description:
            blocks.append(description)

        instruction = getattr(alert, "instruction", None)
        if instruction:
            blocks.append(instruction)

        times: list[str] = []
        sent = getattr(alert, "sent", None)
        if sent is not None:
            times.append(f"Issued: {format_datetime(sent, date_style, time_12h)}")
        expires = getattr(alert, "expires", None)
        if expires is not None:
            times.append(f"Expires: {format_datetime(expires, date_style, time_12h)}")
        if times:
            blocks.append("\n".join(times))

        return "\n\n".join(blocks)

    @staticmethod
    def _copy_payload(alert, settings) -> str:
        """
        Text placed on the clipboard when the Copy button is pressed.

        Identical in both display styles (separate and combined).
        """
        return AlertDialog._build_combined_text(alert, settings)

    def _build_info_text(self) -> str:
        """
        Build the alert info text with severity, urgency, and certainty.

        Note: Areas are intentionally omitted as the Details section
        already contains location information.
        """
        metadata = []
        severity = getattr(self.alert, "severity", None)
        urgency = getattr(self.alert, "urgency", None)
        certainty = getattr(self.alert, "certainty", None)

        if severity:
            metadata.append(f"Severity: {severity}")
        if urgency:
            metadata.append(f"Urgency: {urgency}")
        if certainty:
            metadata.append(f"Certainty: {certainty}")

        return ", ".join(metadata)

    def _setup_accessibility(self):
        """Set up accessibility labels for screen readers."""
        if hasattr(self, "subject_ctrl"):
            self.subject_ctrl.SetName("Subject with alert headline")

        if hasattr(self, "info_ctrl"):
            self.info_ctrl.SetName("Alert information with severity, urgency, and certainty")

        if hasattr(self, "details_ctrl"):
            self.details_ctrl.SetName("Alert details")

        if hasattr(self, "instr_ctrl"):
            self.instr_ctrl.SetName("Instructions")

        if hasattr(self, "combined_ctrl"):
            self.combined_ctrl.SetName("Full alert text")

    def _on_key(self, event):
        """Handle key press - close dialog on Escape."""
        if event.GetKeyCode() == wx.WXK_ESCAPE:
            self.Close()
        else:
            event.Skip()

    def _on_close(self, event):
        """Handle close button press."""
        self.EndModal(wx.ID_CLOSE)

File: rust/crates/aw-app/src/ui/alert_dialog.rs
//! The alert details dialog (`ui/dialogs/alert_dialog.py`) and the summary
//! dialog for several new alerts at once (`ui/dialogs/alerts_summary_dialog.py`).

use std::rc::Rc;

use aw_core::model::{Timestamp, WeatherAlert};
use aw_core::settings::AppSettings;
use wxdragon::prelude::*;

use super::main_window::window;
use crate::app::with_state;

const COPY_LABEL: &str = "Cop&y to clipboard";
const COPY_NAME: &str = "Copy alert text to clipboard";
const CLOSE_LABEL: &str = "&Close";
const SUMMARY_TITLE: &str = "New Weather Alerts";
const SUMMARY_INTRO: &str =
    "Multiple new alerts arrived in this update. Review the summaries below.";
const SUMMARY_NAME: &str = "New alert summaries";
/// The summary dialog's Close button has no mnemonic.
const SUMMARY_CLOSE_LABEL: &str = "Close";
/// wxID_COPY, which wxDragon does not export.
const ID_COPY: i32 = wxdragon::ffi::WXD_ID_COPY as i32;

/// Open the details dialog for `alert` over the main window, with the current
/// settings: `show_alert_dialog(main_window, alert, get_settings())`.
pub(crate) fn show_alert_details(alert: &WeatherAlert) {
    let (Some(w), Some(state)) = (window(), with_state()) else {
        return;
    };
    let settings = state.borrow().config.settings.clone();
    show_alert_dialog(&w.frame, alert, &settings);
}

/// Open the "New Weather Alerts" summary over the main window:
/// `show_alerts_summary_dialog(main_window, alerts)`.
// Called by the immediate alert popups, which the notification port wires up.
#[allow(dead_code)]
pub(crate) fn show_alerts_summary(alerts: &[WeatherAlert]) {
    if let Some(w) = window() {
        show_alerts_summary_dialog(&w.frame, alerts);
    }
}

// ---------------------------------------------------------------------------
// Text builders
// ---------------------------------------------------------------------------

fn non_empty(s: Option<&str>) -> Option<&str> {
    s.filter(|s| !s.is_empty())
}

/// The dialog title. Python formats the attribute as-is, so an alert without
/// an event is titled "Alert: None".
fn dialog_title(alert: &WeatherAlert) -> String {
    format!("Alert: {}", alert.event.as_deref().unwrap_or("None"))
}

/// `_build_subject_text`: the headline, else the event, else "Weather Alert".
fn subject_text(alert: &WeatherAlert) -> &str {
    non_empty(alert.headline.as_deref())
        .or(non_empty(alert.event.as_deref()))
        .unwrap_or("Weather Alert")
}

/// `_build_info_text`: "Severity: X, Urgency: Y, Certainty: Z", skipping empty values.
fn info_text(alert: &WeatherAlert) -> String {
    [
        ("Severity", &alert.severity),
        ("Urgency", &alert.urgency),
        ("Certainty", &alert.certainty),
    ]
    .iter()
    .filter(|(_, v)| !v.is_empty())
    .map(|(k, v)| format!("{k}: {v}"))
    .collect::<Vec<_>>()
    .join(", ")
}

/// `format_datetime` (`display/presentation/time_formatters.py`), in the
/// timestamp's own offset.
fn format_datetime(dt: &Timestamp, date_style: &str, time_12hour: bool) -> String {
    let date_fmt = match date_style {
        "us_short" => "%m/%d/%Y",
        "us_long" => "%B %d, %Y",
        "eu" => "%d/%m/%Y",
        _ => "%Y-%m-%d",
    };
    let time = if time_12hour {
        dt.format("%I:%M %p")
            .to_string()
            .trim_start_matches('0')
            .to_string()
    } else {
        dt.format("%H:%M").to_string()
    };
    format!("{} {time}", dt.format(date_fmt))
}

/// `_build_combined_text`: subject, description, instruction and the
/// Issued/Expires lines as blank-line separated blocks. It is also the
/// clipboard text in both display styles (`_copy_payload`).
fn combined_text(alert: &WeatherAlert, settings: &AppSettings) -> String {
    let fmt = |dt| format_datetime(dt, &settings.date_format, settings.time_format_12hour);
    let mut blocks = vec![subject_text(alert).to_string()];
    blocks.extend(non_empty(Some(alert.description.as_str())).map(str::to_string));
    blocks.extend(non_empty(alert.instruction.as_deref()).map(str::to_string));
    let times: Vec<String> = [("Issued", &alert.sent), ("Expires", &alert.expires)]
        .into_iter()
        .filter_map(|(k, dt)| dt.as_ref().map(|dt| format!("{k}: {}", fmt(dt))))
        .collect();
    if !times.is_empty() {
        blocks.push(times.join("\n"));
    }
    blocks.join("\n\n")
}

/// `AlertsSummaryDialog._build_summary_text`.
fn summary_text(alerts: &[WeatherAlert]) -> String {
    alerts
        .iter()
        .enumerate()
        .map(|(i, alert)| {
            let mut lines = vec![
                format!("{}. {}", i + 1, subject_text(alert)),
                format!("Severity: {}", alert.severity),
            ];
            let optional = [
                ("Urgency", Some(alert.urgency.as_str())),
                ("Certainty", Some(alert.certainty.as_str())),
                ("Details", Some(alert.description.as_str())),
            ];
            for (k, v) in optional {
                if let Some(v) = non_empty(v) {
                    lines.push(format!("{k}: {v}"));
                }
            }
            lines.join("\n")
        })
        .collect::<Vec<_>>()
        .join("\n\n")
}

/// One labelled read-only text field of the alert dialog.
#[derive(Debug, PartialEq)]
struct Field {
    label: &'static str,
    /// `SetName` from `_setup_accessibility`.
    name: &'static str,
    value: String,
    /// Fixed height; `None` means default height, taking the spare room.
    height: Option<i32>,
}

/// `_create_ui`: the fields for the configured `alert_display_style`, in
/// creation order. Separate: Subject, then Alert Info, Details and
/// Instructions when they have text. Combined: one field with everything.
fn alert_fields(alert: &WeatherAlert, settings: &AppSettings) -> Vec<Field> {
    if settings.alert_display_style == "combined" {
        return vec![Field {
            label: "Alert:",
            name: "Full alert text",
            value: combined_text(alert, settings),
            height: None,
        }];
    }
    let mut fields = vec![Field {
        label: "Subject:",
        name: "Subject with alert headline",
        value: subject_text(alert).to_string(),
        height: Some(60),
    }];
    let info = info_text(alert);
    if !info.is_empty() {
        fields.push(Field {
            label: "Alert Info:",
            name: "Alert information with severity, urgency, and certainty",
            value: info,
            height: Some(60),
        });
    }
    if let Some(description) = non_empty(Some(alert.description.as_str())) {
        fields.push(Field {
            label: "Details:",
            name: "Alert details",
            value: description.to_string(),
            height: None,
        });
    }
    if let Some(instruction) = non_empty(alert.instruction.as_deref()) {
        fields.push(Field {
            label: "Instructions:",
            name: "Instructions",
            value: instruction.to_string(),
            height: Some(80),
        });
    }
    fields
}

// ---------------------------------------------------------------------------
// Dialogs
// ---------------------------------------------------------------------------

fn new_dialog(parent: &dyn WxWidget, title: &str, width: i32, height: i32) -> (Dialog, Panel) {
    let dialog = Dialog::builder(parent, title)
        .with_style(DialogStyle::DefaultDialogStyle | DialogStyle::ResizeBorder)
        .with_size(width, height)
        .build();
    let panel = Panel::builder(&dialog).build();
    // Escape closes, like the title bar's close box.
    dialog.bind_internal(EventType::CHAR_HOOK, move |e: Event| {
        if e.get_key_code() == Some(WXK_ESCAPE) {
            e.skip(false);
            dialog.close(false);
        } else {
            e.skip(true);
        }
    });
    (dialog, panel)
}

fn read_only_text(panel: &Panel, value: &str, height: i32) -> TextCtrl {
    TextCtrl::builder(panel)
        .with_value(value)
        .with_style(TextCtrlStyle::MultiLine | TextCtrlStyle::ReadOnly | TextCtrlStyle::Rich2)
        .with_size(Size::new(-1, height))
        .build()
}

fn close_button(dialog: Dialog, panel: &Panel, label: &str) -> Button {
    let button = Button::builder(panel)
        .with_id(ID_CLOSE)
        .with_label(label)
        .build();
    button.on_click(move |_| dialog.end_modal(ID_CLOSE));
    button
}

fn finish_layout(dialog: &Dialog, panel: &Panel, sizer: BoxSizer) {
    panel.set_sizer(sizer, true);
    let dialog_sizer = BoxSizer::builder(Orientation::Vertical).build();
    dialog_sizer.add(panel, 1, SizerFlag::Expand, 0);
    dialog.set_sizer(dialog_sizer, true);
}

/// `show_alert_dialog` / `AlertDialog`.
fn show_alert_dialog(parent: &dyn WxWidget, alert: &WeatherAlert, settings: &AppSettings) {
    let (dialog, panel) = new_dialog(parent, &dialog_title(alert), 700, 500);
    let sizer = BoxSizer::builder(Orientation::Vertical).build();

    let mut first = None;
    for (i, field) in alert_fields(alert, settings).into_iter().enumerate() {
        let label = StaticText::builder(&panel).with_label(field.label).build();
        if let Some(mut font) = label.get_font() {
            font.make_bold();
            label.set_font(&font);
        }
        let top = if i == 0 {
            SizerFlag::Top
        } else {
            SizerFlag::empty()
        };
        sizer.add(&label, 0, SizerFlag::Left | SizerFlag::Right | top, 15);
        let ctrl = read_only_text(&panel, &field.value, field.height.unwrap_or(-1));
        ctrl.set_name(field.name);
        let proportion = i32::from(field.height.is_none());
        sizer.add(
            &ctrl,
            proportion,
            SizerFlag::Expand | SizerFlag::Left | SizerFlag::Right | SizerFlag::Bottom,
            15,
        );
        first.get_or_insert(ctrl);
    }

    let buttons = BoxSizer::builder(Orientation::Horizontal).build();
    buttons.add_stretch_spacer(1);
    let copy = Button::builder(&panel)
        .with_id(ID_COPY)
        .with_label(COPY_LABEL)
        .build();
    copy.set_name(COPY_NAME);
    buttons.add(&copy, 0, SizerFlag::Right, 10);
    let close = close_button(dialog, &panel, CLOSE_LABEL);
    buttons.add(&close, 0, SizerFlag::empty(), 0);
    sizer.add_sizer(&buttons, 0, SizerFlag::Expand | SizerFlag::All, 15);
    finish_layout(&dialog, &panel, sizer);
    if let Some(first) = first {
        first.set_focus();
    }

    // "Copied!" / "Copy failed" shows for two seconds, then the label returns.
    let revert = Rc::new(Timer::new(&dialog));
    revert.on_tick(move |_| {
        copy.set_label(COPY_LABEL);
        panel.layout();
    });
    let payload = combined_text(alert, settings);
    let timer = revert.clone();
    copy.on_click(move |_| {
        let copied = Clipboard::get().set_text(&payload);
        if !copied {
            tracing::warn!("Alert copy: could not open clipboard");
        }
        copy.set_label(if copied { "Copied!" } else { "Copy failed" });
        panel.layout();
        timer.stop();
        timer.start(2000, true);
    });

    dialog.show_modal();
    revert.stop();
    dialog.destroy();
}

/// `show_alerts_summary_dialog` / `AlertsSummaryDialog`.
fn show_alerts_summary_dialog(parent: &dyn WxWidget, alerts: &[WeatherAlert]) {
    let (dialog, panel) = new_dialog(parent, SUMMARY_TITLE, 720, 480);
    let sizer = BoxSizer::builder(Orientation::Vertical).build();
    let intro = StaticText::builder(&panel)
        .with_label(SUMMARY_INTRO)
        .build();
    sizer.add(&intro, 0, SizerFlag::All, 15);
    let summary = read_only_text(&panel, &summary_text(alerts), -1);
    summary.set_name(SUMMARY_NAME);
    sizer.add(
        &summary,
        1,
        SizerFlag::Expand | SizerFlag::Left | SizerFlag::Right | SizerFlag::Bottom,
        15,
    );
    let buttons = BoxSizer::builder(Orientation::Horizontal).build();
    buttons.add_stretch_spacer(1);
    let close = close_button(dialog, &panel, SUMMARY_CLOSE_LABEL);
    buttons.add(&close, 0, SizerFlag::empty(), 0);
    sizer.add_sizer(&buttons, 0, SizerFlag::Expand | SizerFlag::All, 15);
    finish_layout(&dialog, &panel, sizer);
    summary.set_focus();
    dialog.show_modal();
    dialog.destroy();
}

#[cfg(test)]
mod tests {
    //! Golden parity with the Python dialogs, generated by
    //! `rust/tools/golden/alertui.py`.

    use serde_json::{json, Value};

    use super::*;

    fn golden() -> Value {
        serde_json::from_str(include_str!(
            "../../../../testdata/golden/alertui/cases.json"
        ))
        .unwrap()
    }

    fn settings(style: &str, date_format: &str, time_12h: bool) -> AppSettings {
        AppSettings {
            alert_display_style: style.into(),
            date_format: date_format.into(),
            time_format_12hour: time_12h,
            ..AppSettings::default()
        }
    }

    /// The controls `show_alert_dialog` creates, as the generator records them.
    fn controls(fields: &[Field]) -> Vec<Value> {
        let mut out = Vec::new();
        for f in fields {
            out.push(json!({"kind": "label", "label": f.label, "bold": true}));
            out.push(json!({"kind": "text", "name": f.name, "value": f.value}));
        }
        // Python names only the Copy button; "button" is wx's default name.
        out.push(json!({"kind": "button", "id": ID_COPY, "label": COPY_LABEL, "name": COPY_NAME}));
        out.push(json!({"kind": "button", "id": ID_CLOSE, "label": CLOSE_LABEL, "name": "button"}));
        out
    }

    #[test]
    fn alert_dialog_matches_python() {
        let g = golden();
        for case in g["alerts"].as_array().unwrap() {
            let alert: WeatherAlert = serde_json::from_value(case["alert"].clone()).unwrap();
            assert_eq!(dialog_title(&alert), case["title"]);
            assert_eq!(subject_text(&alert), case["subject"]);
            assert_eq!(info_text(&alert), case["info"]);
            for style in ["separate", "combined"] {
                let fields = alert_fields(&alert, &settings(style, "iso", true));
                assert_eq!(
                    Value::from(controls(&fields)),
                    case[format!("{style}_controls")],
                    "{style}: {}",
                    alert.title
                );
                // Initial focus is the first text field.
                assert_eq!(fields[0].name, case[format!("{style}_focus")]);
            }
            for (i, s) in g["settings"].as_array().unwrap().iter().enumerate() {
                let date_format = s["date_format"].as_str().unwrap();
                let time_12h = s["time_format_12hour"].as_bool().unwrap();
                let text = combined_text(&alert, &settings("separate", date_format, time_12h));
                assert_eq!(text, case["combined"][i], "{date_format} {time_12h}");
                assert_eq!(text, case["copy"][i]);
            }
        }
    }

    #[test]
    fn summary_dialog_matches_python() {
        let g = golden();
        let alerts: Vec<WeatherAlert> = g["alerts"]
            .as_array()
            .unwrap()
            .iter()
            .map(|c| serde_json::from_value(c["alert"].clone()).unwrap())
            .collect();
        for case in g["summaries"].as_array().unwrap() {
            let batch: Vec<WeatherAlert> = case["alerts"]
                .as_array()
                .unwrap()
                .iter()
                .map(|i| alerts[i.as_u64().unwrap() as usize].clone())
                .collect();
            let text = summary_text(&batch);
            assert_eq!(text, case["text"]);
            assert_eq!(case["title"], SUMMARY_TITLE);
            assert_eq!(
                case["controls"],
                json!([
                    {"kind": "label", "label": SUMMARY_INTRO, "bold": false},
                    {"kind": "text", "name": SUMMARY_NAME, "value": text},
                    {"kind": "button", "id": ID_CLOSE, "label": SUMMARY_CLOSE_LABEL, "name": "button"},
                ])
            );
        }
    }
}
